src/main/java/com/google/devtools/build/lib/analysis/IncompatiblePlatformProvider.java

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.analysis.platform.ConstraintValueInfo;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import javax.annotation.Nullable;
+
+/**
+ * Provider instance for the {@code target_compatible_with} attribute.
+ *
+ * <p>The presence of this provider is used to indicate that a target is incompatible with the
+ * current platform. Any target that provides this will automatically be excluded from {@link
+ * SkyframeAnalysisResult}'s list of configured targets.
+ *
+ * <p>This provider is able to keep track of _why_ the corresponding target is considered
+ * incompatible. If the target is incompatible because the target platform didn't satisfy one of the
+ * constraints in target_compatible_with, then all the relevant constraints are accessible via
+ * {@code getConstraintsResponsibleForIncompatibility()}. On the other hand, if the corresponding
+ * target is incompatible because one of its dependencies is incompatible, then all the incompatible
+ * dependencies are available via {@code getTargetResponsibleForIncompatibility()}.
+ */
+@Immutable
+public final class IncompatiblePlatformProvider implements TransitiveInfoProvider {
```

Comment:
One more suggestion: if you use Autovalue for this class, it automatically gives you a good implementation of equals and hashCode, and a toString, none of which are currently present.

Reply:
Done.